docs/manual-setup.adoc

```diff
@@ -0,0 +1,152 @@
++
+Later on, if you decide to update `tbtc`, you can run just `install-tbtc.sh`.
++
+Keep in mind, `tbtc` depends on `keep-ecdsa` and `keep-ecdsa` depends on `keep-core` so if you decide to update `keep-ecdsa`, you can run `install-keep-ecdsa.sh` followed by `install-tbtc.sh`.
++
+If you decide to update `keep-core`, you have to run `install-keep-core.sh` followed by `install-keep-ecdsa.sh` followed by `install-tbtc.sh` or just the entire `install.sh` again.
+
+== Run clients
+The above installation script will configure:
+
+- 1 `keep-core` client
+- 3 `keep-ecdsa` clients
+
+To run the `keep-core` client use:
+```
+./run-core-1.sh
+```
+
+It is enough to run one `keep-core` client to generate a group and produce relay entries. Setting up more than one client locally is possible but consumes more resources.
+
+To run `keep-ecdsa` clients use:
+```
+./run-ecdsa-1.sh
+```
+```
+./run-ecdsa-2.sh
+```
+```
+./run-ecdsa-3.sh
+```
+
+There are at least 3 `keep-ecdsa` clients needed to open a keep. Setting up more than three clients locally is possible but consumes more resources.
+
+== Beacon genesis
+
+Before the beacon is able to produce a first relay entry, genesis needs to happen. Genesis triggers the first random beacon group selection.
+
+Genesis should be triggered after `keep-core` client started with:
+```
+cd keep-core
+KEEP_ETHEREUM_PASSWORD="password" ./keep-core --config configs/config.local.1.toml relay genesis
+```
+
+Bonded ECDSA keep factory from `keep-ecdsa` contracts requests for new relay entry to reseed after each signer selection but it is also possible to request for a new relay entry manually with:
+```
+cd keep-core
+KEEP_ETHEREUM_PASSWORD="password" ./keep-core --config configs/config.local.1.toml relay request
+```
+
+== How to interact with the system
+
+You can interact with the system through the tBTC dApp or automated
+end-to-end tests. Before you start interacting, make sure you:
+
+- Installed all system components using `install.sh` script
+- Have a local Geth instance (`run-geth.sh`) working
+- Have local Bitcoin Core and ElectrumX instances (`run-bitcoin.sh`) working
+- Have 1 `keep-core` and 3 `keep-ecdsa` clients up and running
+
+=== Keep Dashboard dApp
+
+To run the Keep Dashboard dApp invoke:
+```
+./run-keep-dashboard.sh
+```
+
+=== tBTC dApp
+
+To run the tBTC dApp against the local Bitcoin network,
+make sure the tBTC dApp is configured to work with the local ElectrumX instance.
+You can configure this in `./tbtc-dapp/src/wrappers/web3.js` file,
+where `TBTC.withConfig` line occurs.
+Then, you can invoke:
+```
+./run-tbtc-dapp.sh
+```
+The application will be available on `http://localhost:3000`.
+
+=== E2E tests
+
+To run the automated end-to-end scenario switch to Node 14.3.0:
+```
+nvm use 14.3.0
+```
+Then invoke:
+```
+./run-e2e-test.sh
```

Comment:
It failed for me with:
```
added 3656 packages from 1303 contributors and audited 2986 packages in 522.177s

45 packages are looking for funding
  run `npm fund` for details

found 1025 vulnerabilities (1024 low, 1 high)
  run `npm audit fix` to fix them, or `npm audit` for details
(node:79641) ExperimentalWarning: Importing JSON modules is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Initial TBTC balance for account 0xD6b0a1CA8f0641B97eFeC0F1eD73d72e58b38FA5 is: 0

Starting deposit number [1]...

Creating new deposit contract with lot size 100000 satoshis...
undefined Error: value out of range (argument="value", value=20, code=INVALID_ARGUMENT, version=bytes/5.0.0-beta.138)
    at Logger.makeError (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/logger/lib/index.js:178:21)
    at Logger.throwError (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/logger/lib/index.js:187:20)
    at Logger.throwArgumentError (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/logger/lib/index.js:190:21)
    at Object.hexZeroPad (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/bytes/lib/index.js:263:16)
    at AddressCoder.decode (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/address.js:34:45)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/array.js:93:31
    at Array.forEach (<anonymous>)
    at Object.unpack (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/array.js:71:12)
    at TupleCoder.decode (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/tuple.js:39:49)
    at AbiCoder.decode (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/abi-coder.js:93:22)
    at ABICoder.decodeParameters (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-eth-abi/src/index.js:322:30)
    at Contract._decodeMethodReturn (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-eth-contract/src/index.js:568:22)
    at Method.outputFormatter (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-eth-contract/src/index.js:924:42)
    at Method.formatOutput (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-core-method/src/index.js:167:54)
    at sendTxCallback (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-core-method/src/index.js:622:33)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-core-requestmanager/src/index.js:180:9
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/index.js:234:9
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/once.js:12:16
    at replenish (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/eachOfLimit.js:61:25)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/eachOfLimit.js:71:9
    at eachLimit (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/eachLimit.js:43:36)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/doLimit.js:9:16 {
  reason: 'value out of range',
  code: 'INVALID_ARGUMENT',
  argument: 'value',
  value: 20,
  baseType: 'address',
  name: null,
  type: 'address'
}
Test errored out with error:  Error: gas required exceeds allowance (231600754) or always failing transaction
    at WebsocketSubprovider._handleSocketMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/subproviders/websocket.js:121:18)
    at WebSocket.onMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/event-target.js:120:16)
    at WebSocket.emit (events.js:315:20)
    at Receiver.receiverOnMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/websocket.js:720:20)
    at Receiver.emit (events.js:315:20)
    at Receiver.dataMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:414:14)
    at Receiver.getData (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:346:17)
    at Receiver.startLoop (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:133:22)
    at Receiver._write (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:69:10)
    at writeOrBuffer (_stream_writable.js:352:12)
    at Receiver.Writable.write (_stream_writable.js:303:10)
    at Socket.socketOnData (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/websocket.js:795:35)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:302:12)
    at readableAddChunk (_stream_readable.js:278:9)
    at Socket.Readable.push (_stream_readable.js:217:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)
```

I had a bunch of `node-gyp` failures along the way but I am not sure if they are related.

Reply:
```
➜  local-setup git:(local-setup-vm) ✗ git submodule
+8f80496654ce027f146fed1252e4bab00029635b keep-core (v1.2.4-rc-343-g8f8049665)
+4e36af3c4ae70b31da2b4cea6b97f9a32c4a89b2 keep-ecdsa (v1.1.0-158-g4e36af3c)
+bc38b184a52bb8028fb747e3cd9cb78ceefda662 relays (solidity/1.1.0-10-gbc38b18)
+eb74f1bc5e8669edd3e07ca53160eccc953ec870 tbtc (v0.14.0-rc-428-geb74f1bc)
+699396de95a721daf2524b4152dac4e2bfc0a422 tbtc-dapp (v0.14.0-rc-73-g699396d)
+f15ffa80e5ce7862c5268501653338db54a9bde8 tbtc.js (v0.12.0-rc-52-gf15ffa8)
```

Reply:
`node-gyp` is unrelated. Regarding the e2e test fail, there is a possibility that one of the ecdsa clients didn't register to the sortition pool. Please do the following:

1. Restart your VM through `vagrant reload` or `vagrant halt && vagrant up`
2. Once it's up, do `vagrant ssh`
3. Check the logs of three ecdsa clients:
```
sudo journalctl -xfu keep-ecdsa-1
sudo journalctl -xfu keep-ecdsa-2
sudo journalctl -xfu keep-ecdsa-3
```
4. If you don't see logs like `operator is not yet registered for application [0x...], waiting` it means everything is good.
5. Rerun e2e tests

Reply:
This paragraph is about manual setup and my ECDSA clients do not live in vagrant, they are deployed locally. There are registered in the sortition pool because I see they are monitoring the status of the application. I tried cleaning up my `node_modules` but I am getting the same error. Isn't it something about a dependency version not locked down properly?

It's weird because it all works as expected on vagrant.

```
(Use `node --trace-warnings ...` to show where the warning was created)
Initial TBTC balance for account 0xD6b0a1CA8f0641B97eFeC0F1eD73d72e58b38FA5 is: 0

Starting deposit number [1]...

Creating new deposit contract with lot size 100000 satoshis...
undefined Error: value out of range (argument="value", value=20, code=INVALID_ARGUMENT, version=bytes/5.0.0-beta.138)
    at Logger.makeError (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/logger/lib/index.js:178:21)
    at Logger.throwError (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/logger/lib/index.js:187:20)
    at Logger.throwArgumentError (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/logger/lib/index.js:190:21)
    at Object.hexZeroPad (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/bytes/lib/index.js:263:16)
    at AddressCoder.decode (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/address.js:34:45)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/array.js:93:31
    at Array.forEach (<anonymous>)
    at Object.unpack (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/array.js:71:12)
    at TupleCoder.decode (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/coders/tuple.js:39:49)
    at AbiCoder.decode (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/@ethersproject/abi/lib/abi-coder.js:93:22)
    at ABICoder.decodeParameters (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-eth-abi/src/index.js:322:30)
    at Contract._decodeMethodReturn (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-eth-contract/src/index.js:568:22)
    at Method.outputFormatter (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-eth-contract/src/index.js:924:42)
    at Method.formatOutput (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-core-method/src/index.js:167:54)
    at sendTxCallback (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-core-method/src/index.js:622:33)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-core-requestmanager/src/index.js:180:9
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/index.js:234:9
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/once.js:12:16
    at replenish (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/eachOfLimit.js:61:25)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/eachOfLimit.js:71:9
    at eachLimit (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/eachLimit.js:43:36)
    at /Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/async/internal/doLimit.js:9:16 {
  reason: 'value out of range',
  code: 'INVALID_ARGUMENT',
  argument: 'value',
  value: 20,
  baseType: 'address',
  name: null,
  type: 'address'
}
Test errored out with error:  Error: gas required exceeds allowance (178421262) or always failing transaction
    at WebsocketSubprovider._handleSocketMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/subproviders/websocket.js:121:18)
    at WebSocket.onMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/event-target.js:120:16)
    at WebSocket.emit (events.js:315:20)
    at Receiver.receiverOnMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/websocket.js:720:20)
    at Receiver.emit (events.js:315:20)
    at Receiver.dataMessage (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:414:14)
    at Receiver.getData (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:346:17)
    at Receiver.startLoop (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:133:22)
    at Receiver._write (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/receiver.js:69:10)
    at writeOrBuffer (_stream_writable.js:352:12)
    at Receiver.Writable.write (_stream_writable.js:303:10)
    at Socket.socketOnData (/Users/piotr/go/src/github.com/keep-network/local-setup/e2e/node_modules/web3-provider-engine/node_modules/ws/lib/websocket.js:795:35)
    at Socket.emit (events.js:315:20)
    at addChunk (_stream_readable.js:302:12)
    at readableAddChunk (_stream_readable.js:278:9)
    at Socket.Readable.push (_stream_readable.js:217:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:186:23)
```

Reply:
First, please add `handleRevert=true` to `TBTC.js` and rerun the tests. Maybe we will get some more specific revert messages from the contract. If not, please try to run a fresh instance of `local-setup` with the latest submodules versions and check if the problem still occurs.

Reply:
I installed keep and tbtc components one more time, cleaning up `node_modules` of `tbtc` and `e2e` and it worked fine this time. I think I could be very unlucky because of some not properly locked dependency in `tbtc` or some temporarily broken dependency. When I tried to reinstall components from scratch I stuck on
```
Installing NPM dependencies...
npm ERR! code ENOLOCAL
npm ERR! Could not install from "node_modules/web3/bignumber.js@git+https:/github.com/debris/bignumber.js.git#94d7146671b9719e00a09c29b01a691bc85048c2" as it does not contain a package.json file.
```
which is obviously not true looking at this commit hash. I had to reset `package-lock.json` and clean up `node_modules` to install `tbtc`. I am not saying that was the reason of my problem for sure, but maybe it was the reason. I can no longer reproduce this error after reinstalling everything 🤷 

I'll mark this conversation as resolved.